In [2]:
import pandas as pd
results = pd.read_csv("../dataset/model_comparison.csv")

results

,Model,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Logistic Regression,0.886610,0.429494,0.905497,0.582634,0.96391
1,Decision Tree,0.999985,0.999943,0.999886,0.999914,0.99994
2,Random Forest,0.999975,1.000000,0.999714,0.999857,1.00000


In [3]:
best_model = results.loc[results["F1 Score"].idxmax()]

best_model


Model        Decision Tree
Accuracy          0.999985
Precision         0.999943
Recall            0.999886
F1 Score          0.999914
ROC-AUC            0.99994
Name: 1, dtype: object

In [4]:
print("Final Model:", best_model["Model"])
print("Accuracy:", round(best_model["Accuracy"], 4))
print("Precision:", round(best_model["Precision"], 4))
print("Recall:", round(best_model["Recall"], 4))
print("F1 Score:", round(best_model["F1 Score"], 4))
print("ROC-AUC:", round(best_model["ROC-AUC"], 4))

Final Model: Decision Tree
Accuracy: 1.0
Precision: 0.9999
Recall: 0.9999
F1 Score: 0.9999
ROC-AUC: 0.9999


In [5]:
print("Key Findings")
print()
print("1. Fraudulent transactions showed strong differences in transaction ratio, distance features and online ordering.")
print("2. Logistic Regression achieved high recall but produced considerably more false positives.")
print("3. Decision Tree and Random Forest performed extremely well on the test dataset.")
print("4. Ratio to median purchase price was the most important feature for the tree-based models.")
print("5. The near-perfect performance is specific to this dataset and may not directly generalize to real-world fraud detection.")

Key Findings

1. Fraudulent transactions showed strong differences in transaction ratio, distance features and online ordering.
2. Logistic Regression achieved high recall but produced considerably more false positives.
3. Decision Tree and Random Forest performed extremely well on the test dataset.
4. Ratio to median purchase price was the most important feature for the tree-based models.
5. The near-perfect performance is specific to this dataset and may not directly generalize to real-world fraud detection.


# Conclusion

This project focused on detecting fraudulent credit card transactions using machine learning.

The dataset contained approximately 1 million transactions with seven input features. Exploratory analysis showed that fraudulent transactions differed significantly from legitimate transactions in features such as ratio to median purchase price, transaction distance and online ordering.

Three classification models were evaluated:

- Logistic Regression
- Decision Tree
- Random Forest

Logistic Regression achieved a strong recall of approximately 90.55%, but its precision was considerably lower. The Decision Tree and Random Forest models performed substantially better on this dataset, achieving almost perfect precision, recall and F1 scores.

The Decision Tree achieved the highest F1 score among the three models, while Random Forest produced a very similar result. Tree-based models also showed that `ratio_to_median_purchase_price` was the most influential feature.

Although the models achieved near-perfect performance, these results should be interpreted carefully. The dataset contains very strong patterns separating fraudulent and legitimate transactions, so the performance may not represent what would be achieved on real-world financial transaction data.

Overall, the project demonstrates the complete machine learning workflow from data exploration and preprocessing to model training, evaluation and threshold analysis.

# Limitations and Future Improvements

- The dataset may not fully represent real-world credit card transactions.
- The extremely high model performance may indicate that the dataset contains unusually strong patterns.
- Real-world fraud detection would require continuous monitoring and model retraining.
- Threshold selection would depend on the business cost of false positives and false negatives.
- Future work could involve testing the models on a more realistic or external dataset.
- Model interpretability and deployment could also be explored in future versions.